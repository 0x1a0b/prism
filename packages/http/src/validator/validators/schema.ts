import { IValidation, ValidationSeverity } from '@stoplight/prism-core';
import { ISchema } from '@stoplight/types';
import * as Ajv from 'ajv';
import { Ajv as AjvClass } from 'ajv';

import { ISchemaValidator } from './types';
import { convertAjvErrors } from './utils';

const SUPPORTED_MEDIATYPES = ['application/json'];

export class JSONSchemaValidator implements ISchemaValidator<ISchema> {
  private ajv: AjvClass;

  constructor() {
    this.ajv = new Ajv({ allErrors: true, messages: true });
  }

  public validate(content: any, schema: ISchema): IValidation[] {
    const validate = this.ajv.compile(schema);
    const errors = validate(JSON.parse(content)) ? [] : validate.errors;
    return convertAjvErrors(errors, ValidationSeverity.ERROR);
  }

  public supports(mediaType: string) {
    return SUPPORTED_MEDIATYPES.indexOf(mediaType) > -1;
  }
}
