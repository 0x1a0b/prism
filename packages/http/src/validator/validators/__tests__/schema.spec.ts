import { JSONSchemaValidator } from '../schema';

describe('JSONSchemaValidator', () => {
  const jsonSchemaValidator = new JSONSchemaValidator();

  describe('supports()', () => {
    describe('supported media type', () => {
      it('returns true', () => {
        expect(jsonSchemaValidator.supports('application/json')).toBe(true);
      });
    });

    describe('unsupported media type', () => {
      it('returns false', () => {
        expect(jsonSchemaValidator.supports('application/x')).toBe(false);
      });
    });
  });

  describe('validate()', () => {
    it('validates positively', () => {
      expect(jsonSchemaValidator.validate('"str"', { type: 'string' })).toEqual([]);
    });

    it('validates negatively', () => {
      expect(jsonSchemaValidator.validate('{"key":"str"}', { type: 'string' })).toMatchSnapshot();
    });
  });
});
