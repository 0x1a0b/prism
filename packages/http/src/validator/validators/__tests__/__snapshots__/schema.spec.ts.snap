// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaValidator validate() validates negatively 1`] = `
Array [
  Object {
    "message": "should be string",
    "name": "type",
    "path": Array [],
    "severity": "error",
    "summary": "should be string",
  },
]
`;
